File: Buddy/buddy-ios/Sources/buddy/Models/Personality.swift
import Foundation

enum PersonalityTrait: String, Codable {
    case carinoso     // mucho pet/play
    case glotton      // mucha comida
    case dormilon     // mucho sueño
    case juguetón     // mucho play
    case neutral

    var emoji: String {
        switch self {
        case .carinoso: "🥰"
        case .glotton:  "🍽"
        case .dormilon: "😴"
        case .juguetón: "🎉"
        case .neutral:  "😐"
        }
    }

    var label: String {
        switch self {
        case .carinoso: "Cariñoso"
        case .glotton:  "Glotón"
        case .dormilon: "Dormilón"
        case .juguetón: "Juguetón"
        case .neutral:  "Neutral"
        }
    }
}

/// Tracks interactions to derive emerging personality.
struct PersonalityTracker: Codable {
    var feedCount: Int = 0
    var playCount: Int = 0
    var petCount: Int = 0
    var sleepCount: Int = 0

    mutating func record(_ action: PetAction) {
        switch action {
        case .eat:   feedCount += 1
        case .play:  playCount += 1; petCount += 1
        case .sleep: sleepCount += 1
        default: break
        }
    }

    var derivedTrait: PersonalityTrait {
        let total = feedCount + playCount + petCount + sleepCount
        guard total > 5 else { return .neutral }
        let counts: [(PersonalityTrait, Int)] = [
            (.glotton, feedCount),
            (.juguetón, playCount),
            (.carinoso, petCount),
            (.dormilon, sleepCount)
        ]
        let max = counts.max(by: { $0.1 < $1.1 })!
        return Double(max.1) / Double(total) > 0.4 ? max.0 : .neutral
    }

    private static let key = "buddy.personality.v1"
    static func load() -> PersonalityTracker {
        guard let data = UserDefaults.standard.data(forKey: key),
              let tracker = try? JSONDecoder().decode(PersonalityTracker.self, from: data) else {
            return PersonalityTracker()
        }
        return tracker
    }
    func save() {
        if let data = try? JSONEncoder().encode(self) {
            UserDefaults.standard.set(data, forKey: Self.key)
        }
    }
    static func reset() {
        UserDefaults.standard.removeObject(forKey: key)
    }
}

File: Buddy/buddy-flutter/lib/models/personality.dart
import 'dart:convert';

import 'package:shared_preferences/shared_preferences.dart';

import 'pet_action.dart';

enum PersonalityTrait {
  carinoso,
  glotton,
  dormilon,
  jugueton,
  neutral;

  String get emoji => switch (this) {
        PersonalityTrait.carinoso => '🥰',
        PersonalityTrait.glotton  => '🍽',
        PersonalityTrait.dormilon => '😴',
        PersonalityTrait.jugueton => '🎉',
        PersonalityTrait.neutral  => '😐',
      };

  String get label => switch (this) {
        PersonalityTrait.carinoso => 'Cariñoso',
        PersonalityTrait.glotton  => 'Glotón',
        PersonalityTrait.dormilon => 'Dormilón',
        PersonalityTrait.jugueton => 'Juguetón',
        PersonalityTrait.neutral  => 'Neutral',
      };
}

class PersonalityTracker {
  int feedCount;
  int playCount;
  int petCount;
  int sleepCount;

  PersonalityTracker({
    this.feedCount = 0,
    this.playCount = 0,
    this.petCount = 0,
    this.sleepCount = 0,
  });

  void record(PetAction action) {
    switch (action) {
      case PetAction.eat:   feedCount++; break;
      case PetAction.play:  playCount++; petCount++; break;
      case PetAction.sleep: sleepCount++; break;
      default: break;
    }
  }

  PersonalityTrait get derivedTrait {
    final total = feedCount + playCount + petCount + sleepCount;
    if (total <= 5) return PersonalityTrait.neutral;
    final counts = <PersonalityTrait, int>{
      PersonalityTrait.glotton:  feedCount,
      PersonalityTrait.jugueton: playCount,
      PersonalityTrait.carinoso: petCount,
      PersonalityTrait.dormilon: sleepCount,
    };
    final maxEntry = counts.entries.reduce((a, b) => a.value >= b.value ? a : b);
    return maxEntry.value / total > 0.4 ? maxEntry.key : PersonalityTrait.neutral;
  }

  Map<String, dynamic> toJson() => {
        'feedCount': feedCount,
        'playCount': playCount,
        'petCount': petCount,
        'sleepCount': sleepCount,
      };

  factory PersonalityTracker.fromJson(Map<String, dynamic> j) => PersonalityTracker(
        feedCount: j['feedCount'] as int? ?? 0,
        playCount: j['playCount'] as int? ?? 0,
        petCount: j['petCount'] as int? ?? 0,
        sleepCount: j['sleepCount'] as int? ?? 0,
      );

  static const _key = 'buddy.personality.v1';

  static Future<PersonalityTracker> load() async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(_key);
    if (raw == null) return PersonalityTracker();
    try {
      return PersonalityTracker.fromJson(jsonDecode(raw) as Map<String, dynamic>);
    } catch (_) {
      return PersonalityTracker();
    }
  }

  Future<void> save() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_key, jsonEncode(toJson()));
  }

  static Future<void> reset() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_key);
  }
}
